Fetch current active and selected programs at Home Connect

Diegorro98 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

At the moment, when the integration is first loaded there's no info about which program is running neither which one is selected.

This PR aims to fetch the active program and the selected program so that it can be used to determine the value of program switch entities and select and active select entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Diegorro98 committed:
Fetch current active and selected programs

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -237,22 +237,54 @@ async def _async_update_data(self) -> dict[str, HomeConnectApplianceData]:
+                                )
+                            ).programs
+                        )
+                    except HomeConnectError as error:
```
Why do we catch `HomeConnectError` here but not above?

MartinHjelmare commented on homeassistant/components/home_connect/select.py:
```diff
@@ -66,18 +66,17 @@ def __init__(
             for program in appliance.programs
             if program.key != ProgramKey.UNKNOWN
         ]
+        super().__init__(
```
Why do we move this?

Diegorro98 replied:
I think that it was a change that I made while the migration to aiohomeconnect was WIP, and it was because initializing `HomeConnectEntity` triggered `async_write_ha_state`, and because options weren't set, it raised an exception.
Any way, it doesn't happen any more, so I'm reverting that change

Diegorro98 committed:
Intialize HomeConnectEntity first at SelectProgramEntity

Diegorro98 committed:
Use the right exception

Diegorro98 committed:
Merge branch 'dev' into home_connect/fetch_current_program

Diegorro98 committed:
Use active/selected program from `get_all_programs`

This will allow us to reduce the number of requests that we need to perform to get all the data ready (only one requests vs. three requests)

Diegorro98 committed:
Remove no longer required mocks

Diegorro98 commented:
moving to draft because possible changes for ArrayOfPrograms model at aiohomeconnect

Diegorro98 committed:
Merge branch 'dev' into home_connect/fetch_current_program

Diegorro98 committed:
Fix

MartinHjelmare reviewed (approved):
Thanks!
